File: py/factory.py
from abc import ABC, abstractmethod

class ITransport(ABC):

  @abstractmethod
  def message(self): ...


class IFactory(ABC):

  @abstractmethod
  def createTransport(self) -> ITransport:
    ...

  def getMessage(self):
    transport = self.createTransport()
    transport.message()


class Machine(ITransport):

  def message(self):
    print("Врум.")


class Bike(ITransport):

  def message(self):
    print("Вжих!")

class MachineFactory(IFactory):

  def createTransport(self) -> ITransport:
    return Machine()


class BikeFactory(IFactory):

  def createTransport(self) -> ITransport:
    return Bike()


def ClientCode(factory: IFactory):
  factory.getMessage()


if __name__ == "__main__":
  ClientCode(MachineFactory())
  ClientCode(BikeFactory())

File: go/factory.go
package main

type IFactory interface {
	GetMessage()
	CreateTransport() ITransport
}

type MachineFactory struct{}

func (t *MachineFactory) CreateTransport() ITransport {
	return &Machine{}
}

func (t *MachineFactory) GetMessage() {
	t.CreateTransport().Message()
}

type BikeFactory struct{}

func (t *BikeFactory) CreateTransport() ITransport {
	return &Bike{}
}

func (t *BikeFactory) GetMessage() {
	t.CreateTransport().Message()
}

type ITransport interface {
	Message()
}

type Machine struct{}
type Bike struct{}

func (t *Machine) Message() {
	print("Врум.\n")
}

func (t *Bike) Message() {
	print("Вжих!\n")
}

func ClientCode(factory IFactory) {
	factory.GetMessage()
}

func main() {
	ClientCode(&MachineFactory{})
	ClientCode(&BikeFactory{})
}
